core/variant/array.cpp

```diff
@@ -233,8 +233,45 @@ void Array::push_back(const Variant &p_value) {
 }
 
 void Array::append_array(const Array &p_array) {
-	ERR_FAIL_COND(!_p->typed.validate(p_array, "append_array"));
-	_p->array.append_array(p_array._p->array);
+	if (_p->typed.type != Variant::OBJECT && _p->typed.type == p_array._p->typed.type) {
+		// Same type or untyped, should be fine
+		_p->array.append_array(p_array._p->array);
+	} else if (_p->typed.type == Variant::NIL) { // From typed to untyped
+		_p->array.append_array(p_array._p->array);
+	} else if (p_array._p->typed.type == Variant::NIL) { // From untyped to typed, must try to check if they are all valid
+		if (_p->typed.type == Variant::OBJECT) {
+			// For objects, it needs full validation, either can be converted or fail
+			for (int i = 0; i < p_array._p->array.size(); i++) {
+				ERR_FAIL_COND(!_p->typed.validate(p_array._p->array[i], "assign"));
+			}
+			_p->array.append_array(p_array._p->array);
+		} else {
+			// For non objects, we need to check if there is a valid conversion, which needs to happen one by one, so this is the worst case.
+			Vector<Variant> new_array;
+			new_array.resize(p_array._p->array.size());
+			for (int i = 0; i < p_array._p->array.size(); i++) {
+				Variant src_val = p_array._p->array[i];
+				if (src_val.get_type() == _p->typed.type) {
+					new_array.write[i] = src_val;
+				} else if (Variant::can_convert_strict(src_val.get_type(), _p->typed.type)) {
```

Comment:
Wouldn't this create an inconsistency:
```gdscript
var int_array: Array[int] = [1, 2, 3]
func _ready():
    int_array.append_array([1.23])     # would be allowed
    int_array.append(1.23)             # still not allowed
    int_array[0] = 1.23                # still not allowed
```

Reply:
I personally am in favor of allowing these assignments to do implicit casting. An inconsistency, yes, but the other operations you've listed I would consider a little restrictive.